File: coinchange2.java
// Time Complexity : O(mn) where m is no. of different denominations
// and n is the total amount.
class Solution {
    public int change(int amount, int[] coins) {
        int m = coins.length;
        int n = amount;
        int [] dp = new int[n+1];
        dp[0] = 1;
        for(int i = 1; i <= m; i++){
            for(int j = 0; j <= n; j++){
                // denomination of coin > amount
                if(coins[i-1] <= j){
                    dp[j] = dp[j] + dp[j - coins[i-1]];
                }
            }
        }
        return dp[n];
    }

    //Test Case 1: amount = 5, coins = [1,2,5]
    //Test Case 2: amount = 3, coins = [2]
    //Test Case 3: amount = 10, coins = [10]
    public static void main(String args[]){
        Solution obj = new Solution();
        int[] coins1= new int[] {1, 2, 5};
        System.out.println(obj.change(5, coins1));
        int[] coins2= new int[] {2};
        System.out.println(obj.change(3, coins2));
        int[] coins3= new int[] {10};
        System.out.println(obj.change(10, coins3));
    }
}

File: coin_change_2.cpp
// Time Complexity: O(n*m)
// Space complexity: O(n)
// Did this code successfully run on Leetcode : Yes

// Approach: Dynamic Programming
// 1. We will create a dp 1 D array.
// 2. We will fill the dp array with 0 at the 0th index.
// 3. Then we will iterate over the coins array and fill the dp array with the number of ways to make the amount with the respective coin.
// 4. We will return the last element of the dp array.

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        vector<int>dp(amount+1, 0);
        dp[0] = 1;
        for(int i = 0; i<coins.size(); i++){
            for(int j = coins[i]; j<= amount; j++){
                dp[j] += dp[j - coins[i]];
            }
        }
        return dp[amount];
    }
};
